File: problems/1834_single_threaded_cpu/Solution.java
import java.util.*;

class Solution {
    public int[] getOrder(int[][] tasks) {
        int n = tasks.length;
        int[][] indexed = new int[n][3];
        for (int i = 0; i < n; i++) { indexed[i] = new int[]{tasks[i][0], tasks[i][1], i}; }
        Arrays.sort(indexed, (a, b) -> a[0] - b[0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] != b[0] ? a[0] - b[0] : a[1] - b[1]);
        int[] result = new int[n];
        long time = 0;
        int i = 0, ri = 0;
        while (i < n || !pq.isEmpty()) {
            if (pq.isEmpty() && i < n && indexed[i][0] > time) time = indexed[i][0];
            while (i < n && indexed[i][0] <= time) { pq.offer(new int[]{indexed[i][1], indexed[i][2]}); i++; }
            int[] top = pq.poll();
            time += top[0];
            result[ri++] = top[1];
        }
        return result;
    }
}

File: problems/1834_single_threaded_cpu/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    vector<int> getOrder(vector<vector<int>>& tasks) {
        int n = tasks.size();
        vector<int> idx(n);
        iota(idx.begin(), idx.end(), 0);
        sort(idx.begin(), idx.end(), [&](int a, int b) { return tasks[a][0] < tasks[b][0]; });
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> pq;
        vector<int> result;
        long long time = 0;
        int i = 0;
        while (i < n || !pq.empty()) {
            if (pq.empty() && i < n && tasks[idx[i]][0] > time) time = tasks[idx[i]][0];
            while (i < n && tasks[idx[i]][0] <= time) { pq.push({tasks[idx[i]][1], idx[i]}); i++; }
            auto [pt, id] = pq.top(); pq.pop();
            time += pt;
            result.push_back(id);
        }
        return result;
    }
};
